src/script_opt/CPP/Stmts.cc

```diff
@@ -369,6 +369,11 @@ void CPPCompile::GenWhenStmt(const WhenInfo* wi, const std::string& when_lambda,
         Emit("\tthrow CPPDelayedCallException();");
         Emit("return %s;", GenericValPtrToGT("retval", ret_type, GEN_NATIVE));
         EndBlock();
+
+        // Return something to avoid return-without-value warnings - but not
+        // if this is a when-inside-a-when, or a hook.
+        if ( ! func_type->ExpressionlessReturnOkay() && ! in_hook )
+            Emit("return 0;");
```

Comment:
It's okay to return 0 because the code will actually never execute. The C++ compiler can't figure that out, however, because the code path is dead (under the circumstances mentioned) due to run-time logic that occurs elsewhere. Basically, this is a `return when` construct, which has its own way of short-circuiting execution which occurs earlier.